pepper-apis/src/main/java/org/broadinstitute/ddp/service/EventService.java

```diff
@@ -0,0 +1,136 @@
+                    .getOrCreateNumOccurrences(eventConfig.getEventConfigurationId(),
+                            eventSignal.getParticipantId());
+
+            logger.info(
+                    "Checking if the occurrences per user ({}) hit the threshold ({}) for event configuration (id = {}), user id = {}",
+                    numOccurrences,
+                    eventConfig.getMaxOccurrencesPerUser(),
+                    eventConfig.getEventConfigurationId(),
+                    eventSignal.getParticipantId()
+            );
+
+            if (eventConfig.getMaxOccurrencesPerUser() != null && numOccurrences >= eventConfig.getMaxOccurrencesPerUser()) {
+                logger.info(
+                        "The number of this event's configuration occurrences for the participant (id = {}) is {}"
+                                + " while the allowed maximum number of occurrences per user is {}, skipping the configuration",
+                        eventSignal.getParticipantId(), numOccurrences, eventConfig.getMaxOccurrencesPerUser()
+                );
+                return;
+            }
+
+            String cancelExpr = eventConfig.getCancelExpression();
+            if (StringUtils.isNotBlank(cancelExpr)) {
+                logger.info("Checking the cancel expression `{}` for the event configuration (id = {})",
+                        cancelExpr, eventConfig.getEventConfigurationId());
+                try {
+                    boolean shouldCancel = pexInterpreter.eval(cancelExpr, handle, eventSignal.getParticipantGuid(),
+                            null);
+                    if (shouldCancel) {
+                        logger.info("Cancel expression `{}` evaluated to TRUE, skipping the configuration", cancelExpr);
+                        return;
+                    }
+                } catch (PexException e) {
+                    throw new DaoException("Error evaluating cancel expression: " + cancelExpr, e);
+                }
+            }
+
+            // Checking if a precondition expression (if exists) evaluates to TRUE
+            String precondExpr = eventConfig.getPreconditionExpression();
+            logger.info(
+                    "Checking the precondition expression `{}` for the event configuration (id = {})",
+                    eventConfig.getPreconditionExpression(),
+                    eventConfig.getEventConfigurationId()
+            );
+            try {
+                if (precondExpr != null && !pexInterpreter.eval(precondExpr, handle,
+                        eventSignal.getParticipantGuid(), null)) {
+                    logger.info("Precondition expression {} evaluated to FALSE, skipping the configuration", precondExpr);
+                    return;
+                }
+            } catch (PexException e) {
+                throw new DaoException("Error evaluating pex expression " + precondExpr, e);
+            }
+
+            if (eventConfig.isTriggered(handle, eventSignal)) {
```

Comment:
Should we do this check at the start of the function? That way we don't have to waste time doing all these checks when we're not even going to run the action.